include/revng/RestructureCFG/ScopeGraphLoggerPass.h

```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+//
+// Copyright rev.ng Labs Srl. See LICENSE.md for details.
+//
+
+#include "llvm/IR/Function.h"
+#include "llvm/Pass.h"
+
+class ScopeGraphLoggerPass : public llvm::FunctionPass {
```

Comment:
I agree that it doesn't have a practical utility at the moment, but this is true only if we have another method of dumping the `ScopeGraph`, which ideally would be using the `DOTGraphTraits`, that at the moment don't work. I can make this a standard executable in the `tests` folder, but it we merge without any support for dumping the graph in dot, this would be anyway very useful for debugging purposes of anything that interacts with a `ScopeGraph`, so I would like to have a pass available in that case... What do you think?

Reply:
I agree that it's good to have something to dump the scope graph that is based on `DOTGraphTraits`, but I don't agree that it should be an LLVM Pass, for at least a couple of reasons.
1. Maintainability: we still have to migrate away from the legacy pass manager, so anything that is implemented as a pass either needs to be implemented twice (one for the legacy and one for the new PM) or will just work with one of the 2 PMs which is even worse. In either case, the version working with the legacy PM will have to be dropped soonish
2. If you're thinking about something for debug purposes it's not good to tie yourself to an LLVM pass. What if you're in the middle of a transformation on the ScopeGraph and you're looking into it in `gdb` and you want to just do `dumpScopeGraph(TheFunction)`? If you have an LLVM pass you can't do that or you have to cook a separate wrapper that instantiates a whole pass manager with a pipeline inside, just for the purpose of running a single pass.

So overall I think that this should go in unittests. And then when you implement the `DOTGraphTraits` you'll also make `dumpScopeGraph` helper function for debug, and you'll use that.

Reply:
Ok, after the discussions we had both here and on mattermost, I made the following changes:
1) I moved the core of the dump algorithm in a dedicated function here https://github.com/revng/revng/blob/57e9b483c79b868ba12fde41644a3ee0aba57651/include/revng/RestructureCFG/ScopeGraphGraphTraits.h#L214. In this way we have the `dumpScopeGraph` function that we can use to debug the `ScopeGraph`
2) The pass to dump the `ScopeGraph` has been moved to the `tests/unit` folder, it is compiled as a standard library https://github.com/revng/revng/blob/57e9b483c79b868ba12fde41644a3ee0aba57651/tests/unit/CMakeLists.txt#L606, so it is not installed. The pass uses the debug function mentioned in point 1) above.
3) The unit tests manually load the `.so` of the pass, and uses it.

Reply:
Makes sense.